xen: remove `stubdomains` #TODO note

SigmaSquadron opened the pull request:
## Description of changes

MiniOS stubdomains are legacy and on their way to being replaced by unikraft. There's no need to work towards them.

## Things done

- [x] No Rebuilds
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

CertainLach reviewed (approved):
Other than that - LGTM.
Can't say anything about functionality, as I haven't used it despite implementing in my PR.

SigmaSquadron commented:
Result of `nixpkgs-review pr 343111` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>qubes-core-vchan-xen</li>
  </ul>
</details>
<details>
  <summary>84 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>collectd</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>docker-machine-kvm2</li>
    <li>gnome-boxes</li>
    <li>guestfs-tools</li>
    <li>htcondor</li>
    <li>libguestfs</li>
    <li>libguestfs-with-appliance</li>
    <li>librenms</li>
    <li>libvirt</li>
    <li>libvirt-glib</li>
    <li>libvirt-glib.dev</li>
    <li>libvirt-glib.devdoc</li>
    <li>libvmi</li>
    <li>libvmi.dev</li>
    <li>libvmi.lib</li>
    <li>mgmt</li>
    <li>minikube</li>
    <li>multipass</li>
    <li>ocamlPackages.ocaml_libvirt</li>
    <li>perl536Packages.SysVirt</li>
    <li>perl536Packages.SysVirt.devdoc</li>
    <li>perl538Packages.SysVirt</li>
    <li>perl538Packages.SysVirt.devdoc</li>
    <li>python311Packages.guestfs</li>
    <li>python311Packages.guestfs.dist</li>
    <li>python311Packages.libvirt</li>
    <li>python311Packages.libvirt.dist</li>
    <li>python312Packages.guestfs</li>
    <li>python312Packages.guestfs.dist</li>
    <li>python312Packages.libvirt</li>
    <li>python312Packages.libvirt.dist</li>
    <li>qemu_xen</li>
    <li>qemu_xen.debug</li>
    <li>qemu_xen.ga</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>rubyPackages.ruby-libvirt</li>
    <li>rubyPackages_3_1.ruby-libvirt</li>
    <li>rubyPackages_3_2.ruby-libvirt</li>
    <li>vagrant</li>
    <li>virt-manager</li>
    <li>virt-manager-qt</li>
    <li>virt-manager.dist</li>
    <li>virt-top</li>
    <li>virt-viewer</li>
    <li>xen</li>
    <li>xen-guest-agent</li>
    <li>xen-slim</li>
    <li>xen-slim.boot</li>
    <li>xen-slim.dev</li>
    <li>xen-slim.doc</li>
    <li>xen-slim.man</li>
    <li>xen.boot</li>
    <li>xen.dev</li>
    <li>xen.doc</li>
    <li>xen.man</li>
    <li>xenPackages.xen_4_17</li>
    <li>xenPackages.xen_4_17-slim</li>
    <li>xenPackages.xen_4_17-slim.boot</li>
    <li>xenPackages.xen_4_17-slim.dev</li>
    <li>xenPackages.xen_4_17-slim.doc</li>
    <li>xenPackages.xen_4_17-slim.man</li>
    <li>xenPackages.xen_4_17.boot</li>
    <li>xenPackages.xen_4_17.dev</li>
    <li>xenPackages.xen_4_17.doc</li>
    <li>xenPackages.xen_4_17.man</li>
    <li>xenPackages.xen_4_18</li>
    <li>xenPackages.xen_4_18-slim</li>
    <li>xenPackages.xen_4_18-slim.boot</li>
    <li>xenPackages.xen_4_18-slim.dev</li>
    <li>xenPackages.xen_4_18-slim.doc</li>
    <li>xenPackages.xen_4_18-slim.man</li>
    <li>xenPackages.xen_4_18.boot</li>
    <li>xenPackages.xen_4_18.dev</li>
    <li>xenPackages.xen_4_18.doc</li>
    <li>xenPackages.xen_4_18.man</li>
  </ul>
</details>

emilazy commented on pkgs/applications/virtualization/xen/generic/default.nix:
```diff
@@ -220,6 +235,56 @@ let
+        };
+    in
+    {
+      gmp = extFile "gmp" ".tar.bz2";
+      lwip = extFile "lwip" ".tar.gz";
+      newlib = extFile "newlib" ".tar.gz";
+      pciutils = extFile "pciutils" ".tar.bz2";
+      polarssl = extFile "polarssl" "-gpl.tgz";
+      tpm_emulator = extFile "tpm_emulator" ".tar.gz";
+      zlib = extFile "zlib" ".tar.gz";
```
Do these absolutely need to be vendored? Obviously I don’t really know what’s going on here but I would much prefer to use the Nixpkgs versions of these if possible, especially if we’re having to patch their build systems again anyway. I guess we’re invoking their own Buildroot type thing to build their own Linux tree here? It’d be really nice if we could at least use our own `.src`s. Happy to hear an explanation of what’s going on here and why this is a dumb request though :)

SigmaSquadron replied:
> Do these absolutely need to be vendored?

They do, unfortunately. MiniOS stubdomains require ancient versions of all of these libraries, and Xen patches them up as needed during the build process. I tried using nixpkgs' `.src`s[^1], but the build naturally failed as the patches Xen applies did not match the new versions' indexes.

> I guess we’re invoking their own Buildroot type thing to build their own Linux tree here?

More accurately, we're sneaking a whole new (non-Linux) kernel build in here.

> Happy to hear an explanation of what’s going on here and why this is a dumb request though

One of Xen's exclusive features is the stubdomain system. They're tiny, lightweight VMs that take away some of dom0's privilege over the system's hardware. You could have a stubdomain that runs the QEMU device model (`xen.qemu-system-i386`), removing the need for running it in dom0, thus improving the system's security by isolating this attack vector from the host.
MiniOS is one of the operating system/microkernel that runs inside a stubdomain, and is what Xen builds by default. One could also use a Linux-based or [MirageOS-based](https://github.com/NixOS/nixpkgs/pull/23955) stubdomain.

I'll try asking on XenDevel if building MiniOS with ancient dependencies is still relevant, and see what they answer.

[^1]: I tried doing this by adding a hacky function that takes the `${source}.src`, calls `tar` to package them using the correct file extensions, and copy them to the right places in the build's working directory.

SigmaSquadron replied:
So, looks like MiniOS is considered legacy, and is being replaced by `unikraft`. We can still build it, but it should at the very least be a disabled-by-default component.

@CertainLach are you sure this is necessary for Qubes, considering ITL uses Linux-based stubdomains?

emilazy replied:
Thanks for the detailed explanation! Having to bundle these old library versions for their bespoke system build system sucks but it sounds like it’s unavoidable at least for now if we want to package MiniOS. However, if it’s deprecated and being replaced, then I might prefer to dodge this ugliness by skipping packaging it altogether in favour of the newer stuff, as long as nothing like Qubes actually needs it.

SigmaSquadron replied:
I'm not sure Qubes uses it. https://github.com/QubesOS/qubes-vmm-xen-stubdom-linux

If there are no objections from lach, I'd be in favour of closing this. On my personal Xen deployment, I use Linux-based stubdoms anyway.

@digitalrayne do you think there's a need for MiniOS-based stubdoms?

CertainLach replied:
> @CertainLach are you sure this is necessary for Qubes, considering ITL uses Linux-based stubdomains?

No, it is not necessary for qubes, I implemented it because I thought it is needed for qubes while it is not, and because there was a comment for stubdom make target.

CertainLach replied:
Still, I believe NixOS Xen needs to be secure OOTB, and stubdom isolation is a good thing, if MiniOS-based stubdoms are only official way to do that - I think it would be better to have it?

CertainLach replied:
I don't know about MiniOS alternatives, correct me if I'm wrong.

SigmaSquadron replied:
> Still, I believe NixOS Xen needs to be secure OOTB, and stubdom isolation is a good thing, if MiniOS-based stubdoms are only official way to do that - I think it would be better to have it?

Linux-based stubdomains are also a valid way to ensure device isolation. Since they're a legacy component, we can assume that Xen itself doesn't want them being used anymore.

emilazy replied:
This uses a TLS library from 2012, an IP stack from 2008, and a Zlib from 2005. All of those projects have had CVEs since those releases. I’d be more inclined to mark MiniOS with `knownVulnerabilities` than to enable it by default for security reasons.

CertainLach replied:
I agree that it is ancient, the only remaining question from me is

> Does Xen even uses it by default when built with stubdom target?

If not and it needs further configuration anyway - then sure, kill it with fire.

SigmaSquadron replied:
> I don't know about MiniOS alternatives, correct me if I'm wrong. Does Xen even uses it by default when built with stubdom target?

All the `stubdom` target builds are the four `.gz` files listed in `passthru`. You can optionally use them to run pre-built stubdomains by calling them in an `xl` configuration.

digitalrayne replied:
@SigmaSquadron - I agree that it makes sense to remove this. I think it is added maintenance burden for a fairly niche use case, I'm not aware of these being used by any default functionality, especially in the "NixOS as Dom0" use case. I also think that @emilazy makes an excellent point around the additional security overhead of carrying this and making it available. If it were ever re-added I absolutely agree it should be marked  with `knownVulnerabilities`.

SigmaSquadron committed:
xen: drop stubdomains TODO note

It has been decided in #343111 that we should not build MiniOS
stubdomains, and instead use Linux-based ones for now.

Once MirageOS or Unikraft is packaged, we can switch to using those
microkernels as stubdomains.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
@OfBorg build xen

SigmaSquadron commented:
Anyone interested in MiniOS may find this branch useful: https://github.com/SigmaSquadron/nixpkgs/tree/xen-stubdomains-legacy

emilazy commented:
Love it when a PR gets smaller :)
